binding/form_mapping.go

```diff
@@ -198,6 +199,14 @@ func setByForm(value reflect.Value, field reflect.StructField, form map[string][
 }
 
 func setWithProperType(val string, value reflect.Value, field reflect.StructField) error {
+	if value.Kind() != reflect.Struct {
+		if u, ok := value.Addr().Interface().(encoding.TextUnmarshaler); ok {
+			return u.UnmarshalText(bytesconv.StringToBytes(val))
+		}
+	}
+	if u, ok := value.Addr().Interface().(BindUnmarshaler); ok {
+		return u.UnmarshalParam(bytesconv.StringToBytes(val))
+	}
```

Comment:
I believe this will not work if you have a custom slice/array type, like:
```go
type MySlice []string

func (s *MySlice) UmarshalText(...) { ... } // or UnmarshalParam
```
Having such types is useful especially for query or path params. You might want to unmarshal `param=a,b,c` into a `MySlice` field. If you rely on Gin's own understanding of slices, you'd need to use `param=a,param=b,param=c`.

I have implemented this functionality in my (kind of) fork. I have:
```go

type ParamUnmarshaler interface {
	UnmarshalParam(text string) error
}

func trySetCustom(val string, value reflect.Value) (isSet bool, err error) {
	switch v := value.Interface().(type) {
	case encoding.TextUnmarshaler:
		return true, v.UnmarshalText([]byte(val))
	case ParamUnmarshaler:
		return true, v.UnmarshalParam(val)
	}

	switch v := value.Addr().Interface().(type) {
	case encoding.TextUnmarshaler:
		return true, v.UnmarshalText([]byte(val))
	case ParamUnmarshaler:
		return true, v.UnmarshalParam(val)
	}

	return false, nil
}
```
and I call it in three places in `tryToSetValue`:
```go
 	switch value.Kind() {
 	case reflect.Slice:
 		if !ok {
 			vs = []string{opt.defaultValue}
 		}
		if ok, err := trySetCustom(vs[0], value); ok || err != nil {
			return ok, err
		}
 		return true, setSlice(vs, value, field)
 	case reflect.Array:
 		if !ok {
 			vs = []string{opt.defaultValue}
 		}
		if ok, err := trySetCustom(vs[0], value); ok || err != nil {
			return ok, err
		}
 		if len(vs) != value.Len() {
 			return false, fmt.Errorf("%q is not valid value for %s", vs, value.Type().String())
 		}
 		return true, setArray(vs, value, field)
 	default:
 		var val string
 		if !ok {
 			val = opt.defaultValue
 		}
 		if len(vs) > 0 {
 			val = vs[0]
 		}
		if ok, err := trySetCustom(val, value); ok || err != nil {
			return ok, err
		}
 		return true, setWithProperType(val, value, field)
 	}
```